src/test_application/OSPMagnum.cpp

```diff
@@ -160,76 +162,178 @@ osp::active::ActiveScene& OSPMagnum::scene_create(std::string && name)
 void testapp::config_controls(OSPMagnum& rOspApp)
 {
     // Configure Controls
+    //Load toml
+    auto data = toml::parse("settings.toml");
+
+    osp::UserInputHandler& rUserInput = rOspApp.get_input_handler();
+    try {
+        for (const auto& [k, v] : data.as_table()) {
+            std::cout << k << std::endl;
+            std::string primary = toml::find(v, "primary").as_string();
+            std::vector<osp::ButtonVarConfig> controls = parse_control(primary);
+
+            std::string secondary = toml::find(v, "secondary").as_string();
+            std::vector<osp::ButtonVarConfig> secondaryKeys = parse_control(secondary);
+
+            controls.insert(controls.end(), secondaryKeys.begin(), secondaryKeys.end());
+
+            bool holdable = toml::find(v, "holdable").as_boolean();
+
+            //Add user input
+            rUserInput.config_register_control(k, holdable, controls);
+
+        }
+    }
+    catch (std::out_of_range oor) {
+        std::cout << oor.what() << std::endl;
+    }
+}
+
+
+//Map for all the keys
+//The tuple is in this order: device, number, and hold/pressed
+using Key_t = OSPMagnum::KeyEvent::Key;
+using Mouse_t = OSPMagnum::MouseEvent::Button;
+using VarOp_t = osp::ButtonVarConfig::VarOperator;
+using VarTrig_t = osp::ButtonVarConfig::VarTrigger;
+
+typedef std::tuple<int, int> button_tuple;
+const std::map<std::string_view, button_tuple, std::less<>> buttonMap = {
+    //Keyboard
+    {"LCtrl", {osp::sc_keyboard, (int)Key_t::LeftCtrl}},
+    {"RCtrl", {osp::sc_keyboard, (int)Key_t::RightCtrl }},
+    {"LShift", {osp::sc_keyboard, (int)Key_t::LeftShift }},
+    {"RShift", {osp::sc_keyboard, (int)Key_t::RightShift }},
+    {"LAlt", {osp::sc_keyboard, (int)Key_t::LeftAlt }},
+    {"RAlt", {osp::sc_keyboard, (int)Key_t::RightAlt }},
+    {"Up", {osp::sc_keyboard, (int)Key_t::Up }},
+    {"Down", {osp::sc_keyboard, (int)Key_t::Down }},
+    {"Left", {osp::sc_keyboard, (int)Key_t::Left }},
```

Comment:
I don't think that these should be getting cast to int.

Why cant you make the button tuple take the key type directly?

Reply:
Is it because of the mouse keys?

Makes me wonder about two possibilities.

1) Should the keyboard and mouse have their own maps, instead of a shared map?
2) Instead of explicitly saying osp::sc_keyboard vs osp::sc_mouse, you could instead use an std::variant, and then determine if it's a keyboard or mouse mapping based on the item held by the variant.

Reply:
`int` is also how `osp::ButtonVarConfig` stores the button, so I was just following what was already done.

Reply:
Sure.

Maybe @Capital-Asterisk or @z-adams can chime in on it.

I just don't see why casting to int is the right approach. Maybe there's a good reason.